Make universal compare do lexicographic comparison on lists

dolio opened the pull request:
Equality still checks length first, because that should be able to short circuit a bit faster.

Addresses #3603

dolio committed:
Make universal compare do lexocographic comparison on lists

- Eq still checks length first, because that works

pchiusano commented on parser-typechecker/src/Unison/Runtime/Machine.hs:
```diff
@@ -2159,7 +2159,7 @@ universalCompare frn = cmpc False
       | Just sl <- maybeUnwrapForeign Rf.listRef fl,
         Just sr <- maybeUnwrapForeign Rf.listRef fr =
-          comparing Sq.length sl sr <> fold (Sq.zipWith (cmpc tyEq) sl sr)
+          fold (Sq.zipWith (cmpc tyEq) sl sr)
```
I think this isn't correct - it will have `[1,2,3]` equal to `[1,2,3,4]`, right? Since the zipWith cuts off at the shorter list.

I think putting the length comparison after the `zipWith` comparison would do it, but could you add a few tests to `builtins.md` transcript?

dolio replied:
Oh right. Good catch.

dolio committed:
Added back length comparison, but preferring lexicographic

- Just zipping the lists will report that any two lists with the same
  prefix are equal. So, instead, compare the lengths if the zip check
  reports equality.
- Added a few tests for list equality to builtins.md

pchiusano reviewed (approved):
Cool. LGTM!
